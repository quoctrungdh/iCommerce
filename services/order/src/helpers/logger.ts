import winston, { format } from 'winston';
import fluentLogger from 'fluent-logger';
import { FLUENTD_HOST } from '../constants';
const { combine, errors, timestamp } = format;

const print = format.printf((info) => {
  const log = `${info.timestamp}:${info.level}: ${info.message}`;

  return info.stack ? `${log}\n${info.stack}` : log;
});

function loggerHandler() {
  const config = {
    host: FLUENTD_HOST,
    port: 24224,
    timeout: 3.0,
    requireAckResponse: true, // Add this option to wait response from Fluentd certainly
  };

  let fluent: any;
  try {
    const fluentTransport = fluentLogger.support.winstonTransport();
    fluent = new fluentTransport('order.service', config);
  } catch (error) {
    console.log(error.message);
  }

  const wlogger = winston.createLogger({
    format: combine(
      errors({ stack: true }), // <-- use errors format
      timestamp(),
      print
    ),
    transports: [fluent, new winston.transports.Console()],
  });

  wlogger.on('finish', () => {
    fluent.sender.end('end', {}, () => {});
  });

  const loggerCombineMsg = (fns: Function) => {
    return (...messages: any) => {
      const combinedMsg = messages.join('|');
      fns(combinedMsg);
    };
  };

  return {
    info: loggerCombineMsg(wlogger.info),
    error: loggerCombineMsg(wlogger.error),
    debug: loggerCombineMsg(wlogger.debug),
  };
}

export default loggerHandler();
