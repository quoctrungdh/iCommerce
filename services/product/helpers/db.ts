import mongoose from 'mongoose'
import { MONGO_SERVER, MONGO_DB } from '../constants';

export async function initDatabase() {
  if(!MONGO_SERVER) {
    throw new Error("Mongo Server is not found")
  }

  if(!MONGO_DB) {
    throw new Error("Mongo DB is not found")
  }

  console.log({ MONGO_SERVER, MONGO_DB })

  mongoose.Promise = global.Promise;

  await mongoose.connect(MONGO_SERVER, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    dbName: MONGO_DB
  })

  // Get the default connection
  const db = mongoose.connection;
  

  // Bind connection to error event (to get notification of connection errors)
  db.on('error', console.error.bind(console, 'MongoDB connection error:'));
}
