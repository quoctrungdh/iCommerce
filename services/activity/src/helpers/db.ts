import mongoose from 'mongoose';
import { MONGO_SERVER, MONGO_DB } from '../constants';
import logger from '../helpers/logger';

export default async function initDatabase() {
  if (!MONGO_SERVER) {
    throw new Error('Mongo Server is not found');
  }

  if (!MONGO_DB) {
    throw new Error('Mongo DB is not found');
  }

  mongoose.Promise = global.Promise;

  const readyStatus = [
    'disconnected',
    'connected',
    'connecting',
    'disconnecting',
  ];

  const dbClient = await mongoose.connect(`${MONGO_SERVER}/${MONGO_DB}`, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
  });

  // Get the default connection
  // const db = dbClient.connection;
  // logger.info('readyState', readyStatus[db.readyState]);

  // db.once('open', (res) => {
  //   logger.info('db connection open', JSON.stringify(res));
  // });

  // // Bind connection to error event (to get notification of connection errors)
  // db.on('error', logger.error);

  return dbClient;
}
