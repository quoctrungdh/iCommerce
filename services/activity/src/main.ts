import Koa from 'koa';
import Router from 'koa-router';
import koaLogger from 'koa-logger';
import json from 'koa-json';
import bodyParser from 'koa-bodyparser';
import { PORT, IS_DEV, IS_TEST } from './constants';
import initDatabase from './helpers/db';
import errorHandler from './middlewares/errorHandler';
import {
  getActivities,
  getActivity,
  createActivity,
  updateActivity,
  deleteActivity,
  ping,
} from './controllers/activity.controller';
import logger from './helpers/logger';
import { serializeError } from 'serialize-error';
import NatsHandler from './events/natsHandler';

const app = new Koa();
const router = new Router();
const nc = new NatsHandler();

router
  .get('/ping', ping)
  .get('/', getActivities)
  .get('/:id', getActivity)
  .post('/', createActivity)
  .put('/', updateActivity)
  .delete('/', deleteActivity);

// Middlewares
app.use(errorHandler);
app.use(json());
app.use(bodyParser());
if (IS_DEV) {
  app.use(koaLogger());
}
app.use(nc.natsMiddleware);

// Routes
app.use(router.routes()).use(router.allowedMethods());

/* centralized error handling: */
app.on('error', (err, ctx) => {
  logger.error(JSON.stringify(serializeError(err)));
});

const server = app.listen(PORT, () => {
  initDatabase();
  if (!IS_TEST) {
    logger.info('Koa started', PORT);
  }
});

export default server;
