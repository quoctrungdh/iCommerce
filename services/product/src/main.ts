import Koa from 'koa';
import Router from 'koa-router';
import koaLogger from 'koa-logger';
import json from 'koa-json';
import bodyParser from 'koa-bodyparser';
import { PORT, IS_DEV, IS_TEST } from './constants';
import initDatabase from './helpers/db';
import {
  getProduct,
  getProducts,
  createProduct,
  updateProduct,
  deleteProduct,
  ping,
} from './controllers/product.controller';
import errorHandler from './middlewares/errorHandler';
import logger from './helpers/logger';
import { serializeError } from 'serialize-error';
import NatsHandler from './events/natsHandler';

const app = new Koa();
const router = new Router();
const nc = new NatsHandler();

router
  .get('/ping', ping)
  .get('/', getProducts)
  .get('/:id', getProduct)
  .post('/', createProduct)
  .put('/', updateProduct)
  .delete('/', deleteProduct);

// Middlewares
app.use(errorHandler);
app.use(json());
app.use(bodyParser());
if (IS_DEV) {
  app.use(koaLogger());
}
app.use(nc.natsMiddleware);

// Routes
app.use(router.routes()).use(router.allowedMethods());

/* centralized error handling: */
app.on('error', (err, ctx) => {
  logger.error(JSON.stringify(serializeError(err)));
});

const server = app.listen(PORT, () => {
  initDatabase();
  if (!IS_TEST) {
    logger.info('Koa started', PORT);
  }
});

export default server;
