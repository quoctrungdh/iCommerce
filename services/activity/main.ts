import Koa from "koa";
import Router from "koa-router";
import logger from "koa-logger";
import json from "koa-json";
import { PORT } from './constants'
import { initDatabase } from "./helpers/db";
import errorHandler from "./middlewares/errorHandler";
import { getActivities, getActivity, createActivity, updateActivity, deleteActivity, ping } from "./controllers/activity.controller";
import natsMiddleware from './middlewares/nats'

const app = new Koa();
const router = new Router();

router.get("/ping", ping);
router.get("/", getActivities);
router.get("/:id", getActivity);
router.post("/", createActivity);
router.put("/", updateActivity);
router.delete("/", deleteActivity);

// Middlewares
app.use(errorHandler);
app.use(json());
app.use(logger());
app.use(natsMiddleware)

// Routes
app.use(router.routes()).use(router.allowedMethods());

/* centralized error handling: */
app.on('error', (err, ctx) => {
  console.log("onerror", err)
  /* 
   *   console.log error
   *   write error to log file
   *   save error and request information to database if ctx.request match condition
   *   ...
  */
});

app.listen(PORT, () => {
  initDatabase()
  console.log("Koa started", PORT);
});