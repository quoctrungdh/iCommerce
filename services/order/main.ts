import Koa from "koa";
import Router from "koa-router";
import koaLogger from "koa-logger";
import json from "koa-json";
import bodyParser from 'koa-bodyparser';
import { PORT, IS_DEV } from './constants'
import { initDatabase } from "./helpers/db";
import errorHandler from "./middlewares/errorHandler";
import { ping, getOrders, getOrder, createOrder, updateOrder, deleteOrder } from "./controllers/order.controller";
import logger from './helpers/logger'
import { serializeError } from 'serialize-error';
import NatsHandler from "./events/natsHandler";

const app = new Koa();
const router = new Router();
const nc = new NatsHandler()

router.get("/ping", ping);
router.get("/", getOrders);
router.get("/:id", getOrder);
router.post("/", createOrder);
router.put("/", updateOrder);
router.delete("/", deleteOrder);

// Middlewares
app.use(errorHandler);
app.use(json());
app.use(bodyParser());
if(IS_DEV) {
  app.use(koaLogger());
}
app.use(nc.natsMiddleware)

// Routes
app.use(router.routes()).use(router.allowedMethods());

/* centralized error handling: */
app.on('error', (err, ctx) => {
  logger.error(JSON.stringify(serializeError(err)))
});

app.listen(PORT, () => {
  initDatabase()
  console.log("Koa started", PORT);
});