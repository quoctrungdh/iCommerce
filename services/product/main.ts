import Koa from "koa";
import Router from "koa-router";
import logger from "koa-logger";
import json from "koa-json";
import { PORT } from './constants'
import { initDatabase } from "./helpers/db";
import { getProduct, getProducts, createProduct, updateProduct, deleteProduct } from "./controllers/product.controller";
import errorHandler from "./middlewares/errorHandler";

const app = new Koa();
const router = new Router();

router.get("/", getProducts);
router.get("/:id", getProduct);
router.post("/", createProduct);
router.put("/", updateProduct);
router.delete("/", deleteProduct);

// Middlewares
app.use(errorHandler);
app.use(json());
app.use(logger());

// Routes
app.use(router.routes()).use(router.allowedMethods());

/* centralized error handling: */
app.on('error', (err, ctx) => {
  console.log("onerror", err)
  /* 
   *   console.log error
   *   write error to log file
   *   save error and request information to database if ctx.request match condition
   *   ...
  */
});

app.listen(PORT, () => {
  initDatabase()
  console.log("Koa started", PORT);
});