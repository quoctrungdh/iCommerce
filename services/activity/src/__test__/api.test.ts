import mockingoose from 'mockingoose';
import '../helpers/sinon-mongoose';
import ActivityModel from '../models/activity.model';
import supertest from 'supertest';
import main from '../main';

jest.mock('ts-nats', () => ({
  connect: () => ({
    on: jest.fn(),
    publish: jest.fn(),
    subscribe: jest.fn(),
  }),
  Payload: {
    JSON: 'json',
  },
}));
jest.mock('winston', () => ({
  format: {
    colorize: jest.fn(),
    combine: jest.fn(),
    label: jest.fn(),
    timestamp: jest.fn(),
    printf: jest.fn(),
    errors: jest.fn(),
  },
  createLogger: jest.fn().mockReturnValue({
    info: jest.fn(),
    error: jest.fn(),
    debug: jest.fn(),
    log: jest.fn(),
    on: jest.fn(),
  }),
  transports: {
    Console: jest.fn(),
  },
}));
jest.mock('fluent-logger', () => ({
  support: {
    winstonTransport: () => jest.fn(),
  },
}));

const _data = [
  {
    _id: '507f191e810c19729de860ea',
    type: 'SEARCH',
    data: '{"limit":"10","page":"0","brand":"OOOP","color":"red"}',
    host: 'api.i-commerce.example',
    userAgent: 'PostmanRuntime/7.26.3',
    ip: '10.42.0.69',
    port: '1234',
    prefix: '/oka',
    protocol: 'http',
    createdAt: '2020-08-22T09:23:57.318Z',
  },
  {
    _id: '5f40f1a39be715243a0217b9',
    type: 'VIEW',
    data: '{"limit":"10","page":"0","brand":"OOOP","color":"red"}',
    host: 'api.i-commerce.example',
    userAgent: 'PostmanRuntime/7.26.3',
    ip: '10.42.0.69',
    port: '1234',
    prefix: '/search',
    protocol: 'http',
    createdAt: '2020-08-22T09:23:57.318Z',
  },
  {
    _id: '5f40f1b0f940e4246e0cd0d0',
    type: 'FILTER',
    data: '{"limit":"10","page":"0","brand":"OOOP","color":"red"}',
    host: 'api.i-commerce.example',
    userAgent: 'PostmanRuntime/7.26.3',
    ip: '10.42.0.69',
    port: '1234',
    prefix: '/filter',
    protocol: 'http',
    createdAt: '2020-08-22T09:23:57.318Z',
  },
];

describe('API test', () => {
  afterAll(() => {
    main.close();
  });

  test('GET /', async () => {
    mockingoose(ActivityModel).toReturn(_data, 'find');
    ActivityModel.find().then((doc) => {
      expect(JSON.parse(JSON.stringify(doc))).toMatchObject(_data);
    });
    const response = await supertest(main).get('/');
    expect(response.status).toEqual(200);
    expect(response.text).toMatchSnapshot();
  });

  test('GET / empty data', async () => {
    mockingoose(ActivityModel).toReturn([], 'find');
    ActivityModel.find().then((doc) => {
      expect(JSON.parse(JSON.stringify(doc))).toMatchObject([]);
    });
    const response = await supertest(main).get('/');
    expect(response.status).toEqual(200);
    expect(response.body).toEqual({ data: [], meta: { page: 0, pageSize: 0 } });
  });

  test('GET /:id', async () => {
    mockingoose(ActivityModel).toReturn(_data[0], 'findOne');
    ActivityModel.findOne({ _id: '5f40f1a39be715243a0217b9' }).then((doc) => {
      expect(JSON.parse(JSON.stringify(doc))).toMatchObject(_data[0]);
    });
    const response = await supertest(main).get('/507f191e810c19729de860ea');
    expect(response.status).toEqual(200);
    expect(response.body).toEqual(_data[0]);
  });

  test('GET /:id - invalid id', async () => {
    mockingoose(ActivityModel).toReturn(null, 'findOne');
    ActivityModel.findOne().then((doc) => {
      expect(JSON.parse(JSON.stringify(doc))).toBeNull();
    });
    const response = await supertest(main).get('/invalid-id-fake');
    expect(response.status).toEqual(404);
  });

  test('POST /', async () => {
    mockingoose(ActivityModel).toReturn(null, 'save');
    ActivityModel.create().then((doc) => {
      expect(doc).toBeUndefined();
    });
  });

  test('PUT /:id', async () => {
    const output = {
      _id: '5f40f42cf4450c2d61d4aa18',
      type: 'VIEW',
      data: '{"limit":"10","page":"0","brand":"OOOP","color":"red"}',
      host: 'api.i-commerce.example',
      userAgent: 'PostmanRuntime/7.26.3',
      ip: '10.42.0.69',
      port: '1234',
      prefix: '/search',
      protocol: 'http',
      createdAt: new Date('2020-08-22T09:23:57.318Z'),
    };
    mockingoose(ActivityModel).toReturn(output, 'findOneAndUpdate');
    ActivityModel.findOneAndUpdate(
      { _id: '5f40f42cf4450c2d61d4aa18' },
      { port: '2345' }
    )
      .exec()
      .then((doc) => {
        expect(doc).toMatchSnapshot();
      });
  });

  test('DELETE /:id', async () => {
    const output = {
      _id: '5f40da5f1fc695c40887b8c1',
      createdAt: new Date('2020-08-22T09:23:57.318Z'),
    };
    mockingoose(ActivityModel).toReturn(output, 'deleteOne');
    ActivityModel.deleteOne({ _id: '5f40da5f1fc695c40887b8c1' })
      .exec()
      .then((doc) => {
        expect(doc).toMatchSnapshot();
      });
  });
});
