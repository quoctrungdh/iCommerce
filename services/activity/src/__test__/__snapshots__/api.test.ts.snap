// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API test DELETE /:id 1`] = `
Object {
  "_id": "5f40da5f1fc695c40887b8c1",
  "createdAt": 2020-08-22T09:23:57.318Z,
}
`;

exports[`API test GET / 1`] = `
"{
  \\"data\\": [
    {
      \\"createdAt\\": \\"2020-08-22T09:23:57.318Z\\",
      \\"_id\\": \\"507f191e810c19729de860ea\\",
      \\"type\\": \\"SEARCH\\",
      \\"data\\": \\"{\\\\\\"limit\\\\\\":\\\\\\"10\\\\\\",\\\\\\"page\\\\\\":\\\\\\"0\\\\\\",\\\\\\"brand\\\\\\":\\\\\\"OOOP\\\\\\",\\\\\\"color\\\\\\":\\\\\\"red\\\\\\"}\\",
      \\"host\\": \\"api.i-commerce.example\\",
      \\"userAgent\\": \\"PostmanRuntime/7.26.3\\",
      \\"ip\\": \\"10.42.0.69\\",
      \\"port\\": \\"1234\\",
      \\"prefix\\": \\"/oka\\",
      \\"protocol\\": \\"http\\"
    },
    {
      \\"createdAt\\": \\"2020-08-22T09:23:57.318Z\\",
      \\"_id\\": \\"5f40f1a39be715243a0217b9\\",
      \\"type\\": \\"VIEW\\",
      \\"data\\": \\"{\\\\\\"limit\\\\\\":\\\\\\"10\\\\\\",\\\\\\"page\\\\\\":\\\\\\"0\\\\\\",\\\\\\"brand\\\\\\":\\\\\\"OOOP\\\\\\",\\\\\\"color\\\\\\":\\\\\\"red\\\\\\"}\\",
      \\"host\\": \\"api.i-commerce.example\\",
      \\"userAgent\\": \\"PostmanRuntime/7.26.3\\",
      \\"ip\\": \\"10.42.0.69\\",
      \\"port\\": \\"1234\\",
      \\"prefix\\": \\"/search\\",
      \\"protocol\\": \\"http\\"
    },
    {
      \\"createdAt\\": \\"2020-08-22T09:23:57.318Z\\",
      \\"_id\\": \\"5f40f1b0f940e4246e0cd0d0\\",
      \\"type\\": \\"FILTER\\",
      \\"data\\": \\"{\\\\\\"limit\\\\\\":\\\\\\"10\\\\\\",\\\\\\"page\\\\\\":\\\\\\"0\\\\\\",\\\\\\"brand\\\\\\":\\\\\\"OOOP\\\\\\",\\\\\\"color\\\\\\":\\\\\\"red\\\\\\"}\\",
      \\"host\\": \\"api.i-commerce.example\\",
      \\"userAgent\\": \\"PostmanRuntime/7.26.3\\",
      \\"ip\\": \\"10.42.0.69\\",
      \\"port\\": \\"1234\\",
      \\"prefix\\": \\"/filter\\",
      \\"protocol\\": \\"http\\"
    }
  ],
  \\"meta\\": {
    \\"page\\": 0,
    \\"pageSize\\": 3
  }
}"
`;

exports[`API test PUT /:id 1`] = `
Object {
  "_id": "5f40f42cf4450c2d61d4aa18",
  "createdAt": 2020-08-22T09:23:57.318Z,
  "data": "{\\"limit\\":\\"10\\",\\"page\\":\\"0\\",\\"brand\\":\\"OOOP\\",\\"color\\":\\"red\\"}",
  "host": "api.i-commerce.example",
  "ip": "10.42.0.69",
  "port": "1234",
  "prefix": "/search",
  "protocol": "http",
  "type": "VIEW",
  "userAgent": "PostmanRuntime/7.26.3",
}
`;
